joint_limits/include/joint_limits/joint_limiter_interface.hpp

```diff
@@ -0,0 +1,181 @@
+        result = false;
+        break;
+      }
+      RCLCPP_INFO(
+        node_logging_itf_->get_logger(), "Limits for joint %zu (%s) are:\n%s", i,
+        joint_names[i].c_str(), joint_limits_[i].to_string().c_str());
+    }
+
+    if (result)
+    {
+      result = on_init();
+    }
+
+    // avoid linters output
+    (void)robot_description_topic;
+
+    return result;
+  }
+
+  /**
+   * Wrapper init method that accepts pointer to the Node.
+   * For details see other init method.
+   */
+  JOINT_LIMITS_PUBLIC virtual bool init(
+    const std::vector<std::string> & joint_names, const rclcpp::Node::SharedPtr & node,
+    const std::string & robot_description_topic = "/robot_description")
+  {
+    return init(
+      joint_names, node->get_node_parameters_interface(), node->get_node_logging_interface(),
+      robot_description_topic);
+  }
+
+  /**
+   * Wrapper init method that accepts pointer to the LifecycleNode.
+   * For details see other init method.
+   */
+  JOINT_LIMITS_PUBLIC virtual bool init(
+    const std::vector<std::string> & joint_names,
+    const rclcpp_lifecycle::LifecycleNode::SharedPtr & lifecycle_node,
+    const std::string & robot_description_topic = "/robot_description")
+  {
+    return init(
+      joint_names, lifecycle_node->get_node_parameters_interface(),
+      lifecycle_node->get_node_logging_interface(), robot_description_topic);
+  }
+
+  JOINT_LIMITS_PUBLIC virtual bool configure(
+    const trajectory_msgs::msg::JointTrajectoryPoint & current_joint_states)
+  {
+    // TODO(destogl): add checks for position
+    return on_configure(current_joint_states);
+  }
+
+  JOINT_LIMITS_PUBLIC virtual bool enforce(
+    trajectory_msgs::msg::JointTrajectoryPoint & current_joint_states,
+    trajectory_msgs::msg::JointTrajectoryPoint & desired_joint_states, const rclcpp::Duration & dt)
+  {
+    // TODO(destogl): add checks if sizes of vectors and number of limits correspond.
+    return on_enforce(current_joint_states, desired_joint_states, dt);
+  }
+
+protected:
+  // Methods that each limiter implementation has to implement
+  JOINT_LIMITS_PUBLIC virtual bool on_init() { return true; }
+
+  JOINT_LIMITS_PUBLIC virtual bool on_configure(
+    const trajectory_msgs::msg::JointTrajectoryPoint & /*current_joint_states*/)
+  {
+    return true;
+  }
+
+  /** \brief Enforce joint limits to desired joint state.
+   *
+   * Filter-specific implementation of the joint limits enforce algorithm.
+   *
+   * \param[in] current_joint_states current joint states a robot is in.
+   * \param[out] desired_joint_states joint state that should be adjusted to obey the limits.
+   * \param[in] dt time delta to calculate missing integrals and derivation in joint limits.
+   */
+  JOINT_LIMITS_PUBLIC virtual bool on_enforce(
+    trajectory_msgs::msg::JointTrajectoryPoint & current_joint_states,
+    trajectory_msgs::msg::JointTrajectoryPoint & desired_joint_states,
+    const rclcpp::Duration & dt) = 0;
```

Comment:
I don't know if using `trajectory_msgs::msg::JointTrajectoryPoint` makes sense here. We are not dealing with any trajectories and semantically it doesn't make sense. We can use `sensor_msgs::msg::JointState`  but only for the current_joint_state

I personally think it makes sense to have a specific struct for this rather than using a non-semantic one.

Reply:
I see your point. I understand that it is right now being used in the JTC. I think having more `on_enforce` methods give more flexibility. I'm in for this, that would be great.

Regarding the custom structure you were mentioning, it would be cool to have something like `JointStateInfo` structure that has methods `getPositions`, `getVelocities`, `getAccelerations`, `getEfforts` and we can use it's pointer as `current_joint_states` and `desired_joint_states`

Whoever is interested in using this JointLimiter, will simply inherit from `JointStateInfo` and implement his own `CustomJointStateInfo` and then implement the above functions, for instance, the `getPositions` from your case would be returning `current_joint_states.positions` or `desired_joint_states.positions`. In this way, we give flexibility to the users, and in your case, we don't have to convert any data.

What do you think?

Reply:
@destogl What do you think about this approach for the future?

Reply:
I don't know about adding any new types here. I think that most of the people will use standard types as `JointTrajectoryPoint` is. Overriding is still possible in the case you are mentioning. One can do the same with `JointTrajectoryPoint`.

Or maybe even better, we need a new type that we define in `control_msgs` that holds everything what we need.

Reply:
@saikishor I have investigated extensively how we can make the type more flexible. I was trying using templates for that, but having two template arguments lead to errors with pluginlib. Therefore, I went back to the old option and added new `using` where we can change type as needed. We could also simply make multiple `enforce` methods that support different types and convert between them internally.

As I don't have a clear idea what a good type would look like, and before I start to overcomplicate implementation, I would say, let's stick with this and see how this will be used in the Resource Manager. This will make us a bit smarter. :)